src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs

```diff
@@ -507,25 +507,23 @@ public void WaitingForDescriptionToChange()
             private IServerSelector DecorateSelector(IServerSelector selector)
             {
                 var settings = _cluster.Settings;
-                if (settings.PreServerSelector != null || settings.PostServerSelector != null)
+                var allSelectors = new List<IServerSelector>();
+
+                if (settings.PreServerSelector != null)
                 {
-                    var allSelectors = new List<IServerSelector>();
-                    if (settings.PreServerSelector != null)
-                    {
-                        allSelectors.Add(settings.PreServerSelector);
-                    }
+                    allSelectors.Add(settings.PreServerSelector);
+                }
 
-                    allSelectors.Add(selector);
+                allSelectors.Add(selector);
 
-                    if (settings.PostServerSelector != null)
-                    {
-                        allSelectors.Add(settings.PostServerSelector);
-                    }
-
-                    return new CompositeServerSelector(allSelectors);
+                if (settings.PostServerSelector != null)
+                {
+                    allSelectors.Add(settings.PostServerSelector);
                 }
 
-                return selector;
+                allSelectors.Add(new LatencyLimitingServerSelector(settings.LocalThreshold));
```

Comment:
Are we now allocating a new `List` and `LatencyLimitingServerSelector` on EVERY server selection?

Should we be concerned about that?

Reply:
It's almost the same logic that we had previously with this configuration: https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver/ClusterRegistry.cs#L91 where we always specified `postServerSelector`. The only difference, that now we will still specify this latency selector even if a user (for some reason) will decide to set `postServerSelector: null`. And this is what the spec requires here (see #5): https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#multi-threaded-or-asynchronous-server-selection

if the application-provided (#4) selector has not been specified, we will still need to call latency selector according to #5.

Reply:
I understand. I wasn't implying this was a bug.

I was pointing out that it looks like we have more memory allocations than before.

Reply:
I don't think that we have additional memory allocation here in most cases at all.
There are only two scenarios where it could happen (but they are really uncommon):
1. When a user specifies `postServerSelector` to `null`.
2. When we create a cluster not through MongoClient like we do here: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Core.TestHelpers/CoreTestConfiguration.cs#L327
These are rare cases and moreover the allocated memory is too small (and this logic won't be called in for example in the endless loop without delay) and as a result, there's no chance this change would somehow affect performance.

Reply:
At the very least we are allocating a new `LatencyLimitingServerSelector` when we did not use to.

But you are probably right that this is not enough to affect performance.

Reply:
The `LatencyLimitingServerSelector` could be allocated once when the `Cluster` is created (add a `_latencyLimitingServerSelector` field to `Cluster`) if we want to avoid having to create a new `LatencyLimitingServerSelector` on EVERY call to `SelectServer`.

Reply:
Done.